File: AdventOfCode.FSharp/2021/Day1.fs
﻿namespace AdventOfCode.FSharp._2021

open AdventOfCode.Core
open AdventOfCode.FSharp
open Xunit

module Day1 =
    let CountIncreases =
        Seq.pairwise
        >> Seq.fold (fun acc (l, r) -> if r > l then acc + 1 else acc) 0

    let Solve = Shared.SplitInt >> CountIncreases

    let SolveWindowed =
        Shared.SplitInt
        >> Seq.windowed 3
        >> Seq.map Seq.sum
        >> CountIncreases

    let exampleInput =
        @"199
200
208
210
200
207
240
269
260
263
"

    [<Fact>]
    let ``Solves Part One Example`` () =
        let actual = Solve exampleInput
        Assert.Equal(7, actual)

    [<Fact>]
    let ``Solves Part Two Example`` () =
        let actual = SolveWindowed exampleInput
        Assert.Equal(5, actual)

type Day1(input: string) =
    interface ISolver with
        member this.SolvePartOne() = Day1.Solve input
        member this.SolvePartTwo() = Day1.SolveWindowed input

File: AdventOfCode.CSharp/2021/Day1.cs
using AdventOfCode.Core;
using Xunit;

namespace AdventOfCode.CSharp._2021;

public class Day1 : Solver {
    private const string ExampleInput =
        @"199
200
208
210
200
207
240
269
260
263
";

    public Day1(string? input = null) : base(input) { }

    public override long SolvePartOne() {
        var increases = 0;
        var lastDepth = default(int?);

        foreach (var depth in Input.SplitInt()) {
            if (depth > lastDepth) ++increases;
            lastDepth = depth;
        }

        return increases;
    }

    public override long SolvePartTwo() {
        var increases = 0;

        var depths = Input.SplitInt();
        for (var index = 3; index < depths.Length; index++)
            if (depths[index] > depths[index - 3])
                ++increases;

        return increases;
    }


    [Fact]
    public void SolvesPart1Example() {
        var solver = new Day1(ExampleInput);
        var actual = solver.SolvePartOne();

        Assert.Equal(7, actual);
    }

    [Fact]
    public void SolvesPart2Example() {
        var solver = new Day1(ExampleInput);
        var actual = solver.SolvePartTwo();

        Assert.Equal(5, actual);
    }
}